SofaKernel/framework/sofa/defaulttype/Mat.h

```diff
@@ -342,17 +370,19 @@ class Mat : public helper::fixed_array<VecNoInit<C,real>,L>
         return m;
     }
 
-    /// Transpose current matrix.
+    /// Transpose the square matrix.
     void transpose()
     {
-        static_assert(L == C, "");
+        static_assert(L == C, "Cannot self-transpose a non-square matrix. Use transposed() instead");
         for (int i=0; i<L; i++)
+        {
             for (int j=i+1; j<C; j++)
             {
                 real t = this->elems[i][j];
```

Comment:
`std::swap( this->elems[i][j], this->elems[j][i] );`

Reply:
Thanks @matthieu-nesme,

What are the benefits of using std::swap, except for code clarity?
I don't know the difference in complexity between std::swap and a manual swap of the variables.
If I look into the stl doc, it seems that, in terms of complexity, std::swap is the same implementation as mine, with the difference that instead of a call to the simple assignement operator, it also calls std::move as it could optimize the assignements in specific cases (specific datatypes, but not in the case of doubles or float values to my understanding)
So it seems to me that in terms of performances my implementation is quite the same.
Though, if I would declare 't' outside of the nested loop, I would save L*(C-1) construction of this variable (modulo compiler optimizations), which I couldn't do with a call to std::swap. (untested assumption... )

So should I correct the code to:
```
real t;
for (int i=0; i<L; i++)
   for (int j=i+1; j<C; j++)
   {
     t = this->elems[i][j];
     this->elems[i][j] = this->elems[j][i];
     this->elems[j][i] = t;
   }

```
or

```
for (int i=0; i<L; i++)
   for (int j=i+1; j<C; j++)
   {
     std::swap(this->elems[i][j], this->elems[j][i]);
   }

```

Reply:
To me std::swap is better for clarity and I trust the STL to provide the best possible implementation regarding the platform so you do not have to worry about it.
if you declare t as const inside the loop, the compiler optimization will allocate it only once and share it anyway (I guess it should also work if it is not const, but you are not helping the compiler!)

Reply:
Ok,
Thanks for your input, I rectify accordingly =)